jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGenerator.java

```diff
@@ -1,196 +1,213 @@
+    private File featuresDir;
+    private File targetDir;
 
-    public AbstractReportConfig createReportConfig( String... args ) {
-        return new AsciiDocReportConfig( args );
+
+    public AbstractReportConfig createReportConfig(String... args) {
+        return new AsciiDocReportConfig(args);
     }
 
+    @Override
     public void generate() {
-        for( ReportModelFile reportModelFile : completeReportModel.getAllReportModels() ) {
-            writeReportModelToFile( reportModelFile.model, reportModelFile.file );
+        if (!prepareDirectories()) {
+            return;
         }
-        generateIndexFile();
-    }
 
-    private void writeReportModelToFile( ReportModel model, File file ) {
-        String targetFileName = Files.getNameWithoutExtension( file.getName() ) + ".asciidoc";
+        completeReportModel.getAllReportModels().stream()
+                .sorted(Comparator.comparing(AsciiDocReportGenerator::byFeatureName))
+                .forEach(this::writeFeatureFile);
 
-        allFiles.add( targetFileName );
-        if( !config.getTargetDir().exists() ) {
-            config.getTargetDir().mkdirs();
-        }
-        File targetFile = new File( config.getTargetDir(), targetFileName );
-        PrintWriter printWriter = PrintWriterUtil.getPrintWriter( targetFile );
+        writeAllScenariosFile();
 
-        try {
-            new AbstractReportModelHandler().handle( model, new AsciiDocReportModelVisitor( printWriter ) );
-        } finally {
-            ResourceUtil.close( printWriter );
-        }
-    }
+        writeFailedScenariosFiled();
 
-    private void generateIndexFile() {
-        PrintWriter printWriter = PrintWriterUtil.getPrintWriter( new File( config.getTargetDir(), "index.asciidoc" ) );
-        try {
-            printWriter.println( "= JGiven Documentation =\n" );
-            printWriter.println( ":toc: left\n" );
-            printWriter.println( "== Scenarios ==\n" );
-            printWriter.println( "=== Classes ===\n" );
-            printWriter.println( "include::allClasses.asciidoc[]" );
-        } finally {
-            ResourceUtil.close( printWriter );
-        }
+        writePendingScenariosFile();
 
-        printWriter = PrintWriterUtil.getPrintWriter( new File( config.getTargetDir(), "allClasses.asciidoc" ) );
-        try {
-            for( String fileName : allFiles ) {
-                printWriter.println( "include::" + fileName + "[]\n" );
-            }
-        } finally {
-            ResourceUtil.close( printWriter );
+        writeStatisticsFile();
 
-        }
+        writeIndexFile(config.getTitle());
 
     }
 
-    class AsciiDocReportModelVisitor implements ReportModelHandler {
-
-        private final PrintWriter writer;
-
-        AsciiDocReportModelVisitor( PrintWriter printWriter ) {
-            this.writer = printWriter;
+    private boolean prepareDirectories() {
+        targetDir = config.getTargetDir();
```

Comment:
do we have a test that`targetDir` won't be `null` before it is used in other methods?

Reply:
No there was no test until now and I'm not sure there is even now. I've added some tests but to be honest the different usage patterns with `generate()`, `setConfig` and `loadReportModel` being public and sometimes called from other classes did confuse me a little bit.

I would have preferred for those to not be public.

I hope that `targetDir` being `null` is now handled properly.